all-sdk/external-client-configuration.md

```diff
@@ -0,0 +1,454 @@
+        upon the client and may change. 💭 Why? We regret not making TLS the default in the past. If TLS is an object,
+        it can have the following possible fields:
+        * `clientCertPath` - File path to mTLS client cert. Mutually exclusive with `clientCertData`.
+        * `clientCertData` - Cert data. Mutually exclusive with `clientCertPath`.
+        * `clientKeyPath` - File path to mTLS client key. Mutually exclusive with `clientKeyData`.
+        * `clientKeyData` - Key data. Mutually exclusive with `clientKeyPath`.
+        * `serverCaCertPath` - File path to server CA cert. Mutually exclusive with `serverCaCertData`.
+        * `serverCaCertData` - CA cert data. Mutually exclusive with `serverCaCertPath`.
+        * `serverName` - Override SNI name when connecting to server.
+        * `disableHostVerification` - Boolean to disable verifying TLS server host. May not be available in all Temporal
+          clients. 💭 Why? Not all SDKs offer this feature today.
+      * `codec` - Remote codec information to use for all encoding/decoding of payloads. May not be available in all
+        Temporal clients. 💭 Why? Not all SDKs allow use of a remote codec today.
+        * `endpoint` - Endpoint to the remote codec.
+        * `auth` - Authorization header for the remote codec.
+      * `grpcMeta` - Object representing HTTP headers. Values must be strings for now (can discuss arrays later).
+        ❓ Should we call this `headers`? We do in some cases and not others.
+  * 💭 Why JSON?
+    * Because YAML and TOML and others require a dependency in our SDKs. AWS for example has a homegrown INI-ish
+      implementation they have had to build in to every language.
+    * We accept that while JSON is technically human-authorable, it's not as easy as other formats. This is acceptable
+      since we will be providing tooling to mutate the config.
+    * ❓ Should we consider "json with comments" and some kind of very basic pre-parser in each language that strips
+      them? This could come later and not part of MVP.
+* This whole format will be defined in proto in the API repo in the `sdk` package.
+  * 💭 Why?
+    * Clear, centralized contract.
+    * Our SDKs already use proto dependency and this gives them stable JSON without requiring separate JSON dependency.
+    * Users can use the proto tooling to read/write configs programmatically since they are just proto JSON.
+* Do not validate that all keys are known in the file.
+  * 💭 Why?
+    * Originally this section was strict validation, but it became clear that we should allow newer-CLI-created config
+      files that may have newer keys to work with older SDKs.
+    * Similarly, it became clear that we're not validating environment variables this way, why config file keys?
+    * We want the tools that mutate this file to have strong validation on the keys being added, so hopefully this
+      avoids the concern of key typos.
```

Comment:
> I don't think we can assume users will use only our tools to mutate config.

It's not a strong assumption, it's just one of the multiple reasons for lack of strict key validation on config load

> What are the consequences of a silent failure (e.g. due to mistyped key), and are we sure that we're not going to add some key in future that might have significant consequences if missing / if a default is inadvertently used?

Depends on the key.

> Also, we should consider providing optional strict validation even if we decide not to do it by default, so users can at least discover problems in a straightforward way.

This adds surface area to loading APIs. Can you suggest the APIs for how this looks? Should we add an option to every way that a config is loaded that fails on unrecognized key just for this use case? Are we sure this feature is worth it on MVP?

Reply:
> > What are the consequences of a silent failure (e.g. due to mistyped key), and are we sure that we're not going to add some key in future that might have significant consequences if missing / if a default is inadvertently used?
> 
> Depends on the key. It's hard to speculate what an older SDK not recognizing a newer config key is because we can't predict what future keys will be added here.

Right, so we can't know if an unknown key is safe to ignore or not, which makes me think that strict validation might actually be quite important. I will preemptively say that I understand we can't protect users against every misconfiguration, but protecting against a mis-typed field name, or recognizing and flagging some unknown config that might be critical, both feel like pretty common issues that can easily be surfaced/avoided.

Also, adding strict validation and relaxing it later is much easier than skipping it now and adding it later. We're much less likely to get complaints about error messages that don't happen.

> > Also, we should consider providing optional strict validation even if we decide not to do it by default, so users can at least discover problems in a straightforward way.
> 
> This adds surface area to loading APIs. Can you suggest the APIs for how this looks? Should we add an option to every way that a config is loaded that fails on unrecognized key just for this use case? Are we sure this feature is worth it on MVP?

Sure, for each `LoadConfig()` function, accept an optional `strict` flag that defaults to `true` (or `DisableStrict` flag that defaults to `false`, if you must, for languages that don't support defaulting to `true`). Though I think I may have talked myself into adding strict validation as part of the MVP, and removing it later if users complain, in which case no new API surface is necessary.

Reply:
I guess it depends on if you want backwards and forwards compatibility with configuration. Most configuration formats I'm aware of are not strict in this way (including AWS), but I don't have a strong enough opinion to continue to resist this, so it will be added.